tests/util/sandbox/test_docker.py

```diff
@@ -0,0 +1,80 @@
+import textwrap
+from typing import Any
+
+import pytest
+
+from inspect_ai import Task, eval, task
+from inspect_ai.dataset import Sample
+from inspect_ai.solver import Generate, Solver, TaskState, solver
+from inspect_ai.util import sandbox
+
+
+@solver
+def exec_solver(exec_args: list[dict[str, Any]]) -> Solver:
+    async def solve(state: TaskState, generate: Generate) -> TaskState:
+        results: list[str] = []
+        for args in exec_args:
+            result = await sandbox().exec(**args)
+            results.append(result.stdout.strip())
+        state.output.completion = "\n".join(results)
+        return state
+
+    return solve
+
+
+@task
+def custom_task(solver: Solver, setup: str | None = None) -> Task:
+    return Task(
+        dataset=[Sample(input="irrelevant", setup=setup)],
+        plan=[solver],
+        sandbox="docker",
+    )
+
+
+def run_eval(task: Task) -> str:
+    result = eval(task, model="mockllm/model")[0]
+    assert result.samples is not None
+    content = result.samples[0].output.choices[0].message.content
+    assert isinstance(content, str)
+    return content
+
+
+@pytest.mark.slow
+def test_docker_sandbox_env() -> None:
```

Comment:
Maybe also add a `@skip_if_no_docker` decorator along the lines of what we have here: https://github.com/UKGovernmentBEIS/inspect_ai/blob/main/tests/test_helpers/utils.py

Reply:
Thanks. Do you also want me to set up docker for the CI run, or just skip in CI?

Reply:
No, you can just skip in CI.

Reply:
I've added this now. Tested a few different scenarios manually:

<details> <summary> run `pytest` from inside a python docker container (docker not installed). Skip reason is "Test doesn't work without Docker installed." </summary>

```python
root@f2d98862ad6e:/inspect_ai# pytest tests/util/sandbox/test_docker.py  -vv
/usr/local/lib/python3.12/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
=================================================================================================================================== test session starts ===================================================================================================================================
platform linux -- Python 3.12.5, pytest-8.3.2, pluggy-1.5.0 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /inspect_ai
configfile: pyproject.toml
plugins: cov-5.0.0, xdist-3.6.1, anyio-4.4.0, dotenv-0.5.2, asyncio-0.24.0
asyncio: mode=Mode.STRICT, default_loop_scope=None
collected 3 items

tests/util/sandbox/test_docker.py::test_docker_sandbox_env SKIPPED (Test doesn't work without Docker installed.)                                                                                                                                                                    [ 33%]
tests/util/sandbox/test_docker.py::test_docker_sandbox_users SKIPPED (Test doesn't work without Docker installed.)                                                                                                                                                                  [ 66%]
tests/util/sandbox/test_docker.py::test_docker_sandbox_nonexistent_user SKIPPED (Test doesn't work without Docker installed.)                                                                                                                                                       [100%]

================================================================================================================================= short test summary info =================================================================================================================================
SKIPPED [1] tests/util/sandbox/test_docker.py:43: Test doesn't work without Docker installed.
SKIPPED [1] tests/util/sandbox/test_docker.py:52: Test doesn't work without Docker installed.
SKIPPED [1] tests/util/sandbox/test_docker.py:71: Test doesn't work without Docker installed.
=================================================================================================================================== 3 skipped in 0.44s ====================================================================================================================================
```

</details>

<details> <summary> Run `pytest` on my mac. Skip reason is "need --runslow option to run" </summary>


```python
❯ pytest tests/util/sandbox/test_docker.py
/Users/rusheb/code/inspect_ai/.venv/lib/python3.12/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
========================================================================================================= test session starts =========================================================================================================
platform darwin -- Python 3.12.5, pytest-8.3.2, pluggy-1.5.0
rootdir: /Users/rusheb/code/inspect_ai
configfile: pyproject.toml
plugins: asyncio-0.24.0, cov-5.0.0, anyio-4.4.0, dotenv-0.5.2, xdist-3.6.1
asyncio: mode=Mode.STRICT, default_loop_scope=None
collected 3 items

tests/util/sandbox/test_docker.py sss                                                                                                                                                                                           [100%]

======================================================================================================= short test summary info =======================================================================================================
SKIPPED [1] tests/util/sandbox/test_docker.py:43: need --runslow option to run
SKIPPED [1] tests/util/sandbox/test_docker.py:52: need --runslow option to run
SKIPPED [1] tests/util/sandbox/test_docker.py:71: need --runslow option to run
========================================================================================================= 3 skipped in 0.66s ==========================================================================================================
```

</details>

<details> <summary> run `pytest --runslow` on my mac. Tests run and pass. </summary>

```
❯ pytest tests/util/sandbox/test_docker.py  --runslow
/Users/rusheb/code/inspect_ai/.venv/lib/python3.12/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
========================================================================================================= test session starts =========================================================================================================
platform darwin -- Python 3.12.5, pytest-8.3.2, pluggy-1.5.0
rootdir: /Users/rusheb/code/inspect_ai
configfile: pyproject.toml
plugins: asyncio-0.24.0, cov-5.0.0, anyio-4.4.0, dotenv-0.5.2, xdist-3.6.1
asyncio: mode=Mode.STRICT, default_loop_scope=None
collected 3 items

tests/util/sandbox/test_docker.py ...                                                                                                                                                                                           [100%]

=============================================================================================================== PASSES ================================================================================================================
_______________________________________________________________________________________________________ test_docker_sandbox_env _______________________________________________________________________________________________________
-------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------

╭─ custom_task (1 sample): mockllm/model ──────────────────────────────────────╮
│ solver: exec_solver                                       dataset: (samples) │
│                                                               scorer: (none) │
│ total time:            0:00:01                                               │
│                                                                              │
│ Log: logs/2024-08-26T14-16-15+01-00_custom-task_CLmUvJWszJB8RjmiG326hv.json  │
╰──────────────────────────────────────────────────────────────────────────────╯
-------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------
 default Skipped - Image is already present locally
______________________________________________________________________________________________________ test_docker_sandbox_users ______________________________________________________________________________________________________
-------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------

╭─ custom_task (1 sample): mockllm/model ──────────────────────────────────────╮
│ solver: exec_solver, setup: #!/bin/bash                   dataset: (samples) │
│         useradd -m myuser                                     scorer: (none) │
│                                                                              │
│ total time:                   0:00:02                                        │
│                                                                              │
│ Log: logs/2024-08-26T14-16-17+01-00_custom-task_khs78uoNtXuUHK6BWvE9nC.json  │
╰──────────────────────────────────────────────────────────────────────────────╯
-------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------
 default Skipped - Image is already present locally
________________________________________________________________________________________________ test_docker_sandbox_nonexistent_user _________________________________________________________________________________________________
-------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------

╭─ custom_task (1 sample): mockllm/model ──────────────────────────────────────╮
│ solver: exec_solver                                       dataset: (samples) │
│                                                               scorer: (none) │
│ total time:            0:00:01                                               │
│                                                                              │
│ Log: logs/2024-08-26T14-16-19+01-00_custom-task_SeR3DG9tMGmu4sBaNArxsC.json  │
╰──────────────────────────────────────────────────────────────────────────────╯
-------------------------------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------------------------------
 default Skipped - Image is already present locally
======================================================================================================= short test summary info =======================================================================================================
PASSED tests/util/sandbox/test_docker.py::test_docker_sandbox_env
PASSED tests/util/sandbox/test_docker.py::test_docker_sandbox_users
PASSED tests/util/sandbox/test_docker.py::test_docker_sandbox_nonexistent_user
========================================================================================================== 3 passed in 6.23s ==========================================================================================================
```

</details>